File: frontend/features/workflows/registry.ts
import {
  bugTriageInputSchema,
  bugTriageOutputSchema,
  bugTriageSampleInput,
  bugTriageSampleOutput,
} from "@/features/workflows/bug-triage/schema";
import {
  meetingActionsInputSchema,
  meetingActionsOutputSchema,
  meetingActionsSampleInput,
  meetingActionsSampleOutput,
} from "@/features/workflows/meeting-actions/schema";
import {
  statusUpdateInputSchema,
  statusUpdateOutputSchema,
  statusUpdateSampleInput,
  statusUpdateSampleOutput,
} from "@/features/workflows/status-update/schema";
import type { WorkflowDefinition, WorkflowId } from "@/features/workflows/types";

export const workflows = [
  {
    id: "bug-triage",
    title: "Bug / Issue Triage",
    shortTitle: "Bug Triage",
    category: "Technical",
    description:
      "Turn scattered reports and known evidence into a reviewable first-pass triage brief.",
    benefit: "Move from symptom collection to a focused investigation plan.",
    problem:
      "Issue reports often mix facts, assumptions, missing evidence, and suggested fixes in one hard-to-scan thread.",
    ctaLabel: "Run demo triage",
    icon: "bug",
    tone: "indigo",
    inputSchema: bugTriageInputSchema,
    outputSchema: bugTriageOutputSchema,
    sampleInput: bugTriageSampleInput,
    sampleOutput: bugTriageSampleOutput,
    inputPreview: ["Observed behavior", "Known evidence", "Expected behavior"],
    resultPreview: ["Confirmed facts", "Evidence gaps", "Recommended checks"],
  },
  {
    id: "meeting-actions",
    title: "Meeting → Action Items",
    shortTitle: "Meeting Actions",
    category: "Collaboration",
    description:
      "Convert working notes into decisions, owned follow-ups, and unresolved questions.",
    benefit: "Leave the meeting with work people can actually pick up.",
    problem:
      "The useful commitments in a meeting are easy to lose inside chronological notes and conversational context.",
    ctaLabel: "Run demo extraction",
    icon: "meeting",
    tone: "cyan",
    inputSchema: meetingActionsInputSchema,
    outputSchema: meetingActionsOutputSchema,
    sampleInput: meetingActionsSampleInput,
    sampleOutput: meetingActionsSampleOutput,
    inputPreview: ["Meeting notes", "Participants", "Date"],
    resultPreview: ["Decisions", "Action items", "Open questions"],
  },
  {
    id: "status-update",
    title: "Work → Status Update",
    shortTitle: "Status Update",
    category: "Operations",
    description:
      "Shape rough progress notes into a concise update matched to the audience and format.",
    benefit: "Share progress without rebuilding the same update every day.",
    problem:
      "Useful delivery context lives across fragments, while every audience expects a different level of detail.",
    ctaLabel: "Run demo update",
    icon: "status",
    tone: "amber",
    inputSchema: statusUpdateInputSchema,
    outputSchema: statusUpdateOutputSchema,
    sampleInput: statusUpdateSampleInput,
    sampleOutput: statusUpdateSampleOutput,
    inputPreview: ["Rough work notes", "Audience", "Update format"],
    resultPreview: ["Completed", "In progress", "Next steps"],
  },
] as const satisfies readonly WorkflowDefinition[];

export function isWorkflowId(value: string): value is WorkflowId {
  return workflows.some((workflow) => workflow.id === value);
}

export function getWorkflow(id: WorkflowId) {
  return workflows.find((workflow) => workflow.id === id)!;
}

File: backend/workflows/registry.py
from dataclasses import dataclass

from ninja import Schema

from workflows.schemas import (
    BugTriageInput,
    BugTriageOutput,
    MeetingActionsInput,
    MeetingActionsOutput,
    StatusUpdateInput,
    StatusUpdateOutput,
)

type WorkflowId = str


@dataclass(frozen=True)
class WorkflowDefinition:
    id: WorkflowId
    title: str
    short_title: str
    category: str
    description: str
    benefit: str
    input_schema: type[Schema]
    output_schema: type[Schema]
    prompt_instructions: str
    prompt_version: str = "v3-team-routing"

    def metadata(self) -> dict[str, str]:
        return {
            "id": self.id,
            "title": self.title,
            "shortTitle": self.short_title,
            "category": self.category,
            "description": self.description,
            "benefit": self.benefit,
            "promptVersion": self.prompt_version,
        }


WORKFLOW_REGISTRY: dict[WorkflowId, WorkflowDefinition] = {
    "bug-triage": WorkflowDefinition(
        id="bug-triage",
        title="Bug / Issue Triage",
        short_title="Bug Triage",
        category="Technical",
        description="Turn scattered reports and known evidence into a reviewable triage brief.",
        benefit="Move from symptom collection to a focused investigation plan.",
        input_schema=BugTriageInput,
        output_schema=BugTriageOutput,
        prompt_instructions=(
            "Create an evidence-bound issue triage brief. Never state a root cause as confirmed "
            "unless the supplied evidence proves it. Classify the report as product-defect, "
            "configuration-or-process, or needs-more-evidence. Route product defects to "
            "engineering, configuration or settings issues to support, and unclear reports to "
            "operations. Separate "
            "confirmed facts from gaps, make checks specific and safe, and only echo a supplied "
            "triageOwnerId; never invent a collaborator identifier."
        ),
    ),
    "meeting-actions": WorkflowDefinition(
        id="meeting-actions",
        title="Meeting → Action Items",
        short_title="Meeting Actions",
        category="Collaboration",
        description="Convert working notes into decisions, follow-ups, and unresolved questions.",
        benefit="Leave the meeting with work people can actually pick up.",
        input_schema=MeetingActionsInput,
        output_schema=MeetingActionsOutput,
        prompt_instructions=(
            "Extract decisions and action items only from the supplied notes. Set an owner or "
            "deadline only when the notes explicitly support it; otherwise use null. Preserve "
            "unresolved ambiguity as an open question or unresolved item. Only echo a supplied "
            "coordinatorId as followUpCoordinatorId; never invent a collaborator identifier."
        ),
    ),
    "status-update": WorkflowDefinition(
        id="status-update",
        title="Work → Status Update",
        short_title="Status Update",
        category="Operations",
        description="Shape rough progress notes into an audience-appropriate status update.",
        benefit="Share progress without rebuilding the same update every day.",
        input_schema=StatusUpdateInput,
        output_schema=StatusUpdateOutput,
        prompt_instructions=(
            "Transform the supplied notes into an accurate status update for the requested "
            "audience and format. Never describe ongoing work as completed. Keep blockers and "
            "next steps distinct, and make the shareable update concise. Only echo a supplied "
            "authorId; never invent a collaborator identifier."
        ),
    ),
}


def get_workflow(workflow_id: str) -> WorkflowDefinition | None:
    return WORKFLOW_REGISTRY.get(workflow_id)
